model/test_results.go

```diff
@@ -181,79 +183,112 @@ func (t *TestResults) Append(ctx context.Context, results []TestResult) error {
 
 	}
 
-	return errors.Wrap(t.appendTestNames(ctx, testNames), "problem updating log metadata during upload")
+	return nil
 }
 
-func (t *TestResults) appendTestNames(ctx context.Context, testNames []string) error {
+// Download returns a TestResult slice with the corresponding results stored in
+// the offline blob storage. The TestResults should be populated and the
+// environment should not be nil.
+func (t *TestResults) Download(ctx context.Context) ([]TestResult, error) {
+	if !t.populated {
+		return []TestResult{}, errors.New("cannot download with populated test results")
+	}
 	if t.env == nil {
-		return errors.New("cannot append with a nil environment")
+		return []TestResult{}, errors.New("cannot download test results with a nil environment")
 	}
 
 	if t.ID == "" {
 		t.ID = t.Info.ID()
 	}
 
-	set := bson.M{}
-	for _, testName := range testNames {
-		set[bsonutil.GetDottedKeyName(testResultsArtifactKey, testResultsArtifactInfoTestsKey, testName)] = false
+	conf := &CedarConfig{}
+	conf.Setup(t.env)
+	if err := conf.Find(); err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem getting application configuration")
 	}
-	updateResult, err := t.env.GetDB().Collection(testResultsCollection).UpdateOne(
+
+	bucket, err := t.Artifact.Type.Create(
 		ctx,
-		bson.M{"_id": t.ID},
-		bson.M{"$set": set},
+		t.env,
+		conf.Bucket.TestResultsBucket,
+		t.Artifact.Prefix,
+		string(pail.S3PermissionsPrivate),
+		false,
 	)
-	grip.DebugWhen(err == nil, message.Fields{
-		"collection":   testResultsCollection,
-		"id":           t.ID,
-		"updateResult": updateResult,
-		"test_names":   testNames,
-		"op":           "append test names to a test results record",
-	})
-	if err == nil && updateResult.MatchedCount == 0 {
-		err = errors.Errorf("could not find test results record with id %s in the database", t.ID)
+	if err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem creating bucket")
+	}
+
+	iter, err := bucket.List(ctx, "")
+	if err != nil {
+		return []TestResult{}, errors.Wrap(err, "problem listing bucket items")
+	}
+
+	results := []TestResult{}
+	for iter.Next(ctx) {
```

Comment:
I suspect that we may want to return some kind of iterator, rather than reading all of the results into memory at once. I'm not sure that we need to do this right now, and this logic does work. 

An iterator around a buffered channel might be good, or at least having a basic iterator type that we could try (and race?) different implementations, even if we start with this one.

Reply:
Yeah given that there will may be 10s of thousands of results per task that is probably a good idea. Definitely something we can push off until I get to the rest code.

Reply:
open a ticket to track